### PR TITLE
Simplify pdf image output.

### PR DESCRIPTION
**Pull request by anntzer**
- Inline _unpack into its only caller writeImages.
- Let _writeImg infer the image size and grayscaleness from the data
  shape.

preliminary work towards https://github.com/matplotlib/matplotlib/issues/15165.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backends/backend_pdf.py [hunk 1]
```diff
@@ -1455,26 +1431,26 @@ def _writePng(self, data):
                 buffer.seek(length, 1)
             buffer.seek(4, 1)   # skip CRC
 
-    def _writeImg(self, data, height, width, grayscale, id, smask=None):
+    def _writeImg(self, data, id, smask=None):
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_pdf.py]: do we ever actually pass in a grayscale image?

Comment (anntzer) [reply lib/matplotlib/backends/backend_pdf.py]: We do pass single-channel images (for alpha).  As for actual grayscale I don't know (see https://github.com/matplotlib/matplotlib/pull/12871 for the similar thing in postscript), but this PR doesn't change that behavior anyways.

**Comment by timhoffm** [comment]
What is the benefit of inlining? To me this looks more cluttered now. The original `_unpack` seemed like a reasonable abstraction layer.

**Comment by anntzer** [comment]
For me this made #15193 easier to write (even though strictly speaking it is orthogonal) -- as things stand it is a bit annoying to have to go through multiple places to track the shapes of the arrays involved.
Although I guess strictly speaking I could do without the inlining of _unpack, and just get rid of width, height and greyscale as arguments to _writeImg.

**Comment by timhoffm** [comment]
I would be happier with that :smiley:.

**Comment by anntzer** [comment]
sure, done

File: lib/matplotlib/backends/backend_pdf.py [hunk 2]
```diff
@@ -1412,14 +1412,14 @@ def imageObject(self, image):
 
     def _unpack(self, im):
         """
-        Unpack the image object im into height, width, data, alpha,
-        where data and alpha are HxWx3 (RGB) or HxWx1 (grayscale or alpha)
-        arrays, except alpha is None if the image is fully opaque.
+        Unpack image array *img* into ``(data, alpha)``, which have shape
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/backends/backend_pdf.py]: fixed

**Pull request merged**